gr-dtv/lib/dvbt2/dvbt2_paprtr_cc_impl.cc

```diff
@@ -808,22 +736,26 @@ int dvbt2_paprtr_cc_impl::work(int noutput_items,
                             }
                             alpha = a;
                             volk_32f_s32f_multiply_32f(
-                                (float*)rNew, (float*)v, alpha, N_TR * 2);
-                            volk_32f_x2_subtract_32f(
-                                (float*)rNew, (float*)r, (float*)rNew, N_TR * 2);
+                                (float*)rNew.data(), (float*)v.data(), alpha, N_TR * 2);
+                            volk_32f_x2_subtract_32f((float*)rNew.data(),
+                                                     (float*)r.data(),
+                                                     (float*)rNew.data(),
+                                                     N_TR * 2);
                         }
                         for (int n = 0; n < papr_fft_size; n++) {
                             ones_freq[(n + m) % papr_fft_size] = ones_time[n];
                         }
                         result = u * alpha;
                         volk_32fc_s32fc_multiply_32fc(
-                            ctemp, ones_freq, result, papr_fft_size);
-                        volk_32f_x2_subtract_32f(
-                            (float*)c, (float*)c, (float*)ctemp, papr_fft_size * 2);
-                        memcpy(r, rNew, sizeof(gr_complex) * N_TR);
+                            ctemp.data(), ones_freq.data(), result, papr_fft_size);
+                        volk_32f_x2_subtract_32f((float*)c.data(),
+                                                 (float*)c.data(),
+                                                 (float*)ctemp.data(),
+                                                 papr_fft_size * 2);
+                        std::copy(std::begin(rNew), std::end(rNew), std::begin(r));
```

Comment:
So `std::copy` translates to a `memcpy`? Or does it incur more overhead? I've seen some `memcpy` in the other files, that might want to be replaced as well.

Reply:
`std::copy` is like `memcpy` but actually knows what it's copying. `memcpy` has to be told how many bytes, which makes it easy to get something wrong, as types change and when copy-pasting.

Usually template implementations actually compile down to faster than simple functions (because the compiler has type access . Though `memcpy` is special in most compilers.

tl;dr: they'll be just as fast, and with today's compilers may even compile down to the exact same instructions. That's my default unless benchmarks prove otherwise.